feat: added missing examples and plugged test gaps

Aditya-ranjan-16 opened the pull request:
### Description

added missing examples and covered test gaps 

[GIT Issue](https://github.ibm.com/GoldenEye/issues/issues/11645)

### Release required?
<!--- Identify the type of release. For information about the changes in a semantic versioning release, see [Release versioning](https://terraform-ibm-modules.github.io/documentation/#/versioning). --->

- [x] No release
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

adds missing examples and covers test gaps

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

Aditya-ranjan-16 commented:
/run pipeline

Aditya-ranjan-16 commented:
/run pipeline

Aditya-ranjan-16 commented:
/run pipeline

Aditya-ranjan-16 commented:
/run pipeline

Aditya-ranjan-16 commented:
pipeline failing by the error
```
Error: StandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ Error: [ERROR] Error while creating key ring : kp.Error: correlation_id='65d51186-fb41-4c44-8de6-344e9bef892e', msg='This operation cannot be completed. Please try your request at a later time. If the issue persists, please contact IBM KeyProtect.'
TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ 
TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │   with module.kms[0].module.kms_key_rings["elasticsearch-key-ring"].ibm_kms_key_rings.key_ring,
TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │   on .terraform/modules/kms.kms_key_rings/main.tf line 5, in resource "ibm_kms_key_rings" "key_ring":
TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │    5: resource "ibm_kms_key_rings" "key_ring" {
TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ 
TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ ---
TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ id: terraform-46882f4c
Error: StandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ summary: '[ERROR] Error while creating key ring : kp.Error:
TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ correlation_id=''65d51186-fb41-4c44-8de6-344e9bef892e'',
TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │   msg=''This operation cannot be completed. Please try your request at a later time.
TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │   If the issue persists, please contact IBM KeyProtect.'''
TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ severity: error
TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ resource: ibm_kms_key_rings
TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ operation: create
TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ component:
TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │   name: github.com/IBM-Cloud/terraform-provider-ibm
TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │   version: 1.74.0
TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ ---
TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ 
TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: ╵
```
re-running the pipeline

Aditya-ranjan-16 commented:
/run pipeline

shemau commented:
> pipeline failing by the error
> 
> ```
> Error: StandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ Error: [ERROR] Error while creating key ring : kp.Error: correlation_id='65d51186-fb41-4c44-8de6-344e9bef892e', msg='This operation cannot be completed. Please try your request at a later time. If the issue persists, please contact IBM KeyProtect.'
> TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ 
> TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │   with module.kms[0].module.kms_key_rings["elasticsearch-key-ring"].ibm_kms_key_rings.key_ring,
> TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │   on .terraform/modules/kms.kms_key_rings/main.tf line 5, in resource "ibm_kms_key_rings" "key_ring":
> TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │    5: resource "ibm_kms_key_rings" "key_ring" {
> TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ 
> TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ ---
> TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ id: terraform-46882f4c
> Error: StandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ summary: '[ERROR] Error while creating key ring : kp.Error:
> TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ correlation_id=''65d51186-fb41-4c44-8de6-344e9bef892e'',
> TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │   msg=''This operation cannot be completed. Please try your request at a later time.
> TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │   If the issue persists, please contact IBM KeyProtect.'''
> TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ severity: error
> TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ resource: ibm_kms_key_rings
> TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ operation: create
> TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ component:
> TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │   name: github.com/IBM-Cloud/terraform-provider-ibm
> TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │   version: 1.74.0
> TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ ---
> TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: │ 
> TestRunStandardUpgradeSolution 2025-01-27T11:43:36Z logger.go:66: ╵
> ```
> 
> re-running the pipeline

This was because the KMS (us-south HPCS) was being upgraded/migrated at the time the test ran.

shemau commented on examples/backup-restore/main.tf:
```diff
@@ -0,0 +1,28 @@
+  resource_group_name          = var.resource_group == null ? "${var.prefix}-resource-group" : null
+  existing_resource_group_name = var.resource_group
+}
+
```
There should be one empty line (not two).

Aditya-ranjan-16 replied:
fixed https://github.com/terraform-ibm-modules/terraform-ibm-icd-elasticsearch/pull/384/commits/5179634350d06be2503826be43e5f414bd87111d

shemau commented on examples/backup-restore/variables.tf:
```diff
@@ -0,0 +1,46 @@
+  description = "Version of the elasticsearch instance. If no value passed, the current ICD preferred version is used."
+  default     = null
+}
+variable "resource_group" {
```
There should be a line break between variables.

Aditya-ranjan-16 replied:
fixed https://github.com/terraform-ibm-modules/terraform-ibm-icd-elasticsearch/commit/5179634350d06be2503826be43e5f414bd87111d

shemau commented on examples/backup-restore/variables.tf:
```diff
@@ -0,0 +1,46 @@
+  default     = []
+}
+
+variable "elasticsearch_db_backup_crn" {
```
Consistency, this is the only backup-restore example (I checked etcd and mongodb) that references the original instance as a backup.

Is the plan to push this example, asis across all ICD modules? In which case maybe it falls into the existing nomiculture and should be 'existing_database_crn'.

Aditya-ranjan-16 replied:
I reviewed the [MongoDB](https://github.com/terraform-ibm-modules/terraform-ibm-icd-mongodb/blob/main/examples/backup-restore/main.tf) and [etcd](https://github.com/terraform-ibm-modules/terraform-ibm-icd-etcd/blob/main/examples/backup-restore/main.tf) backup examples and found them to be similar to my implementation. Could you please help me understand the differences? I also suggest that we standardize this example across all ICD modules. I have addressed the nomenclature to ensure consistency. Looking forward to your insights.

shemau commented on examples/backup-restore/variables.tf:
```diff
@@ -0,0 +1,46 @@
+
+variable "elasticsearch_db_backup_crn" {
+  type        = string
+  description = "The existing CRN of a backup resource to restore from. If null then it will create a new instance first and then create another instance pointing to the backup of the first instance."
```
That is not what the code does.

The code looks up the provided CRN and restores from a backup of that CRN.

Aditya-ranjan-16 replied:
fixed https://github.com/terraform-ibm-modules/terraform-ibm-icd-elasticsearch/commit/5179634350d06be2503826be43e5f414bd87111d

shemau commented on solutions/standard/main.tf:
```diff
@@ -254,7 +254,7 @@ module "es_instance_crn_parser" {
 # Existing instance local vars
 locals {
-  existing_elasticsearch_guid   = var.existing_db_instance_crn != null ? module.es_instance_crn_parser[0].guid : null
+  existing_elasticsearch_guid   = var.existing_db_instance_crn != null ? module.es_instance_crn_parser[0].service_instance : null
```
I have investigated, but do not understand this change. The other changes are fixing tests, but that this is changing source code. The module version, for common-utilities has not changed, so the output should not be changing.

Unless this was a code fix found during testing. Is that why a new test for existing code path was added?

Aditya-ranjan-16 replied:
This was a code fix: `guid` was not a valid key in the object returned by the crn parser, as the code was most likely never tested with a value passed for `existing_db_instance_crn`. Found this bug while adding the test : `TestRunExistingInstance`

shemau commented on tests/pr_test.go:
```diff
@@ -188,6 +199,81 @@ func TestRunBasicExample(t *testing.T) {
 	assert.NotNil(t, output, "Expected some output")
 }
 
+func TestRunExistingInstance(t *testing.T) {
```
Returning to the running too many tests theme. This appears to be adding a new ICD deployment via running another instance of the basic example. 3 PR test cycles a day, across 8 database types is 168 instances per week, just for this test.

In order to facilitate this, is it possible to remove an existing test because it is sufficiently covered by the new test?

To justify the scale, personally I have 6 new PRs since yesterday; and 9 PR reviews plus renovate modules, major-modules, DA and CI pipelines.

Aditya-ranjan-16 replied:
After an internal discussion, we have agreed to remove the following two tests: `TestRunBasicExample` and `TestRunFSCloudExample`.

- The basic example is already being tested in the newly added test `TestRunExistingInstance`
- FSCloud is already being tested in DA test 

Please let me know if this change looks good to you or if you have any suggestions.

shemau reviewed (changes requested):
At a higher level, it should be unnecessary to include common-dev-assets in a PR. This should be bumped on a standalone cycle and by including it, there is a risk of merge conflicts as the main branch changes or potentially rolling it back.

It is good to see backup-restore coverage being added to elasticsearch. However, when compared to etcd and mongodb example it is different. The ICD module tests should be converging, not diverging with every commit.

The basic example changes are good. Converging the source, so that it is more consistent across modules. This reduces maintenance moving forward, allow for cut-n-paste or possibly cherry picking changes from one module to another.

Aditya-ranjan-16 commented:
/run pipeline

Aditya-ranjan-16 commented:
/run pipeline

Aditya-ranjan-16 commented:
/run pipeline

Aditya-ranjan-16 commented:
> At a higher level, it should be unnecessary to include common-dev-assets in a PR. This should be bumped on a standalone cycle and by including it, there is a risk of merge conflicts as the main branch changes or potentially rolling it back.
> 
> It is good to see backup-restore coverage being added to elasticsearch. However, when compared to etcd and mongodb example it is different. The ICD module tests should be converging, not diverging with every commit.
> 
> The basic example changes are good. Converging the source, so that it is more consistent across modules. This reduces maintenance moving forward, allow for cut-n-paste or possibly cherry picking changes from one module to another.

The main branch `common-dev-assets` was last updated a long time back 3 months ago , and it doesn't include the latest elastic search permanent CRN I added recently . hence I updated `common-dev-assets` to point to a latest commit in my branch 

Not sure why this is the case , but I don't see any recent renovate PR to update CI depnedecies

shemau commented:
/run pipeline

shemau reviewed (dismissed):
LGTM

I will follow up as to why renovate is not bumping common-dev-assets automatically.

Aditya-ranjan-16 commented:
/run pipeline

Aditya-ranjan-16 commented:
/run pipeline

Aditya-ranjan-16 commented:
/run pipeline

Aditya-ranjan-16 commented:
/run pipeline

shemau reviewed (approved):
LGTM. All issues addressed.

ocofaigh commented:
@Aditya-ranjan-16 FYI, there are no code changes to the module or DA, just the example and test - so this means we do no need a new release. See [Release versioning](https://terraform-ibm-modules.github.io/documentation/#/versioning?id=release-versioning) for more info

terraform-ibm-modules-ops commented:
:tada: This PR is included in version 1.28.2 :tada:

The release is available on:
- [GitHub release](https://github.com/terraform-ibm-modules/terraform-ibm-icd-elasticsearch/releases/tag/v1.28.2)
- `v1.28.2`

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
